File: src/python/comportamentais/mediator.py
# -*- coding: utf-8 -*-
# sempre que você quer evitar dependência direta entre objetos que colaboram
from typing import Optional


class Mediator:
    def __init__(self):
        self.textbox: Optional[TextBox] = None
        self.checkbox: Optional[CheckBox] = None
        self.button: Optional[Button] = None

    def notify(self, sender: object, event: str) -> None:
        if event in ("text_changed", "checkbox_toggled"):
            if self.textbox and self.checkbox and self.button:
                self.button.enabled = bool(self.textbox.text and self.checkbox.checked)

        elif event == "button_clicked":
            if self.textbox and self.checkbox:
                if self.textbox.text and self.checkbox.checked:
                    print("Enviado com sucesso!")
                else:
                    print("Não é possível enviar. Preencha todos os campos.")


class TextBox:
    def __init__(self, mediator: Mediator):
        self.mediator = mediator
        self.text: str = ""

    def input(self, text: str) -> None:
        self.text = text
        self.mediator.notify(self, "text_changed")


class CheckBox:
    def __init__(self, mediator: Mediator):
        self.mediator = mediator
        self.checked: bool = False

    def toggle(self) -> None:
        self.checked = not self.checked
        self.mediator.notify(self, "checkbox_toggled")


class Button:
    def __init__(self, mediator: Mediator):
        self.mediator = mediator
        self.enabled: bool = False

    def click(self) -> None:
        self.mediator.notify(self, "button_clicked")


# Uso
mediator = Mediator()
textbox = TextBox(mediator)
checkbox = CheckBox(mediator)
button = Button(mediator)

mediator.textbox = textbox
mediator.checkbox = checkbox
mediator.button = button

# Simulação
textbox.input("Olá mundo")
checkbox.toggle()
button.click()  # Enviado com sucesso!

File: src/cpp/comportamentais/mediator/Mediator.cpp
#include "Mediator.hpp"

#include <iostream>

// Mediator
void Mediator::setTextBox(std::shared_ptr<TextBox> textbox) {
    textbox_ = textbox;
}

void Mediator::setCheckBox(std::shared_ptr<CheckBox> checkbox) {
    checkbox_ = checkbox;
}

void Mediator::setButton(std::shared_ptr<Button> button) {
    button_ = button;
}

void Mediator::notify(Component* sender, const std::string& event) {
    auto textbox = textbox_.lock();
    auto checkbox = checkbox_.lock();
    auto button = button_.lock();

    if (!textbox || !checkbox || !button) return;

    if (event == "text_changed" || event == "checkbox_toggled") {
        bool enable = !textbox->getText().empty() && checkbox->isChecked();
        button->setEnabled(enable);
    } else if (event == "button_clicked") {
        if (!textbox->getText().empty() && checkbox->isChecked()) {
            std::cout << "Enviado com sucesso!\n";
        } else {
            std::cout << "Não é possível enviar. Preencha todos os campos.\n";
        }
    }
}

// TextBox
void TextBox::setMediator(std::shared_ptr<Mediator> mediator) {
    mediator_ = mediator;
}

void TextBox::input(const std::string& text) {
    text_ = text;
    if (auto med = mediator_.lock()) med->notify(this, "text_changed");
}

std::string TextBox::getText() const {
    return text_;
}

// CheckBox
void CheckBox::setMediator(std::shared_ptr<Mediator> mediator) {
    mediator_ = mediator;
}

void CheckBox::toggle() {
    checked_ = !checked_;
    if (auto med = mediator_.lock()) med->notify(this, "checkbox_toggled");
}

bool CheckBox::isChecked() const {
    return checked_;
}

// Button
void Button::setMediator(std::shared_ptr<Mediator> mediator) {
    mediator_ = mediator;
}

void Button::click() {
    if (auto med = mediator_.lock()) med->notify(this, "button_clicked");
}

void Button::setEnabled(bool enabled) {
    enabled_ = enabled;
}

bool Button::isEnabled() const {
    return enabled_;
}
